posydon/popsyn/GRB.py

```diff
@@ -0,0 +1,119 @@
+        """Compute GRB beaming factor."""
+        # TODO: check out Github copilot suggestions: 
+        # 'Fong+15', 'Ghirlanda+16', 'Lamb+19'?
+        
+        df[f'S{i}_f_beaming'] = [np.nan] * df.shape[0]
+        if isinstance(GRB_beaming, float):
+            df.loc[sel, f'S{i}_f_beaming'] = GRB_beaming
+        elif GRB_beaming == 'Goldstein+15':
+            n = df.loc[sel,f'S{i}_f_beaming'].shape[0]
+            # generate a opening angle from a lognormal distribution 
+            # as in Table 3 of https://arxiv.org/pdf/1512.04464.pdf
+            theta = 10**(np.random.normal(0.77,0.37,n))/180.*np.pi # radian
+            # compute the solid angle, factor 2 to accounts for the two emispheres
+            solid_angle = 4*np.pi*(1.-np.cos(theta)) 
+            df.loc[sel,f'S{i}_f_beaming'] = solid_angle/(4*np.pi)
+        elif GRB_beaming == 'Lloyd-Ronning+19':
+            if z is None:
+                raise ValueError('Redshift not specified!')
+            else:
+                z = df.loc[sel,'z_formation']
+            n = df.loc[sel,f'S{i}_f_beaming'].shape[0]
+            # generate a opening angle from a lognormal distribution
+            # as in Table 3 of https://arxiv.org/pdf/1512.04464.pdf
+            theta =10**(np.random.normal(0.77*(1+z)**(-0.75),0.37,n))/180.*np.pi # radian
+            # compute the solid angle, factor 2 to accounts for the two emispheres
+            solid_angle = 4*np.pi*(1.-np.cos(theta))
+            df.loc[sel,f'S{i}_f_beaming'] = solid_angle/(4*np.pi)
+        elif GRB_beaming == 'Pescalli+15':
+            def f_B(E,eta):
+                xi=0.5
+                t=25.
+                C = (7e46)**xi
+                fb = np.ones(len(E))
+                fb[E>0.] = (C*(eta*E[E>0.]/t)**(-xi))**(1./(1-xi))
+                return fb
+            E = df.loc[sel,f'S{i}_E_GRB']
+            eta = df.loc[sel,f'S{i}_eta']
+            df.loc[sel,f'S{i}_f_beaming'] = f_B(E,eta)
+        else:
+            raise ValueError(f'GRB beaming factor {GRB_beaming} not supported!')
+    
+    def compute_E_GRB(i, sel):
+        """Compute the GRB energy."""
+        df[f'S{i}_E_GRB'] = np.zeros(df.shape[0])
+        m_disk_radiated = df.loc[sel,f'S{i}_m_disk_radiated']
+        eta = df.loc[sel,f'S{i}_eta']
+        df.loc[sel,f'S{i}_E_GRB'] = eta * m_disk_radiated * Msun * clight**2 # erg
+        
+    def compute_E_GRB_iso(i, sel):
+        """Compute the GRB isotropic equivalent energy."""
+        df[f'S{i}_E_GRB_iso'] = np.zeros(df.shape[0])
+        E_GRB = df.loc[sel,f'S{i}_E_GRB']
+        f_beaming = df.loc[sel,f'S{i}_f_beaming']
+        df.loc[sel,f'S{i}_E_GRB_iso'] = E_GRB/f_beaming # erg
+        
+    def compute_L_GRB_iso(i, sel):
+        """Compute the GRB isotropic equivalet luminosity."""
+        df[f'S{i}_L_GRB_iso'] = np.zeros(df.shape[0])
+        # compute the GRB isotropic equivalent luminosity assuming a
+        # constant average timescale of 2.5s 
```

Comment:
Could you give a reference for the 2.5s? Pescalli+15 had a t=25s, which is a factor 10 different.
May implement it as a free parameter for that function, where you set the default, e.g. to the 2.5s.

Reply:
I don't remember where I took it, probably Bromberg, I am updating to 25s for the moment and citing Pescalli+15. Again all of this needs to be tested and improved by someone doing an active science project.